### PR TITLE
Add MobileNet v1

### LINKED ISSUE #238: MobileNet v1 model
It would be great to have a version of MobileNet based on the [keras implementation](https://github.com/keras-team/keras-applications/blob/master/keras_applications/mobilenet.py).

Paper: [MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications](https://arxiv.org/abs/1704.04861)

### PR DESCRIPTION
**Pull request by texasmichelle**
Fixes #238.

Add MobileNet v1 model based on the [keras implementation](https://github.com/keras-team/keras-applications/blob/master/keras_applications/mobilenet.py).

- [X] MobileNet v1 model
- [X] MobileNet v1 test
- [x] Conditional zero padding
- [x] Dropout probability
- [x] Depth multiplier with minimum depth
- [x] Verify layer architecture with python keras implementation for additional input sizes
- [x] Verify tensor size fed into the GlobalAvgPool2D layer

Follow-on PRs will address creating a dataset (#237) and example that uses this model (#239).

### CODE REVIEW

**Comment by texasmichelle** [comment]
Looking for some initial feedback, esp. around:
1. Which additional function arguments would be useful?
2. What other functionality is missing?

Style & polish suggestions encouraged!

File: Models/ImageClassification/MobileNet.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import TensorFlow
+
+// Original Paper:
+// MobileNets: Efficient Convolutional Neural Networks for Mobile Vision
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Minor style suggestion: I'd recommend placing the paper title in quotes, which might help make it easier to parse with the text wrapping in the comments.

File: Models/ImageClassification/MobileNet.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  @differentiable
+  public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+    // TODO(michellecasbon): Add conditional zero padding
+    let convolved1 = relu6(input.sequenced(through: dConv, batchNorm1))
+    let convolved2 = relu6(convolved1.sequenced(through: conv, batchNorm2))
+    return convolved2
+  }
+
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Minor formatting issue: this extra empty line can probably be removed.

File: Models/ImageClassification/MobileNet.swift (hunk 1 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: lol - I just removed the quotes because I was looking at SqueezeNet and trying to conform to the repo standard.

File: Models/ImageClassification/MobileNet.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  public var dConvBlock7 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 512, strides: (1, 1))
+  public var dConvBlock8 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 512, strides: (1, 1))
+  public var dConvBlock9 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 512, strides: (1, 1))
+  public var dConvBlock10 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 512, strides: (1, 1))
+  public var dConvBlock11 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 512, strides: (1, 1))
+  public var dConvBlock12 = DepthwiseConvBlock(filterCount: 512, pointwiseFilterCount: 1024, strides: (2, 2))
+  public var dConvBlock13 = DepthwiseConvBlock(filterCount: 1024, pointwiseFilterCount: 1024, strides: (1, 1))
+  public var avgPool = GlobalAvgPool2D<Float>()
+  public var reshape = Reshape<Float>(shape: [1, 1, 1, 1024])
+  public var dropout = Dropout<Float>(probability: 0.001)
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: In the Keras model, dropout probability is an input parameter to the model. Do we want to do the same here?

File: Models/ImageClassification/MobileNet.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  public var convLast: Conv2D<Float>
+
+  public init(classCount: Int) {
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: In line with the above comment, in the paper and Keras implementations, a depth multiplier (and minimum depth) are specified as inputs to the model and then used to scale the number of filters to shrink or expand the model. Did we want to match that functionality? If so, the various convolution blocks would probably have to be initialized here with the respective calculated filter counts, rather than above.

File: Models/ImageClassification/MobileNet.swift (hunk 3 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Done.

File: Models/ImageClassification/MobileNet.swift [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  public var dConv: DepthwiseConv2D<Float>
+  public var batchNorm1: BatchNorm<Float>
+  public var conv: Conv2D<Float>
+  public var batchNorm2: BatchNorm<Float>
+
+  public init(filterCount: Int, pointwiseFilterCount: Int, depthMultiplier: Int, strides: (Int, Int)) {
+    if depthMultiplier > 0 {
+        self.depthMultiplier = depthMultiplier
+    } else {
+        print("Depth multiplier must be an integer greater than 0. Setting to default value of 1.")
```
Comment (texasmichelle) [inline Models/ImageClassification/MobileNet.swift]: Is there a better way to output this log message?

File: Models/ImageClassification/MobileNet.swift (hunk 4 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: I'm not sure that I applied the multiplier in the correct place. It was a bit hard to tell since in python it's an argument to DepthwiseConv2D, which is missing in the swift layer. Can you confirm?

File: Models/ImageClassification/MobileNet.swift (hunk 5 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Would printError from ModelSupport be better here? I ran into linking issues running the test command, so prob just didn't import it properly.

**Comment by texasmichelle** [comment]
Is there a preferred linter? This is ready for review.

File: Models/ImageClassification/MobileNet.swift [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+      print(convolved.shape)
+      print("Relu6")
+    }
+    return relu6(convolved)
+  }
+}
+
+public struct DepthwiseConvBlock: Layer {
+  @noDerivative
+  var depthMultiplier: Int
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: If this isn't a learned or otherwise mutated parameter, this should probably be a `let`. Likewise for `strides` and `printSummary`.

File: Models/ImageClassification/MobileNet.swift [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+  @noDerivative
+  var strides: (Int, Int)
+  @noDerivative
+  var printSummary: Bool
+
+  public var zeroPad = ZeroPadding2D<Float>(padding: ((0, 1), (0, 1)))
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Will the zero padding by mutated in any way by training? If not, it can probably be a let.

File: Models/ImageClassification/MobileNet.swift [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    }
+    return convolved3
+  }
+}
+
+public struct MobileNetV1: Layer {
+  @noDerivative
+  var classCount: Int
+  @noDerivative
+  var printSummary: Bool
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Similarly, `classCount` and `printSummary` can probably be lets here.

File: Models/ImageClassification/MobileNet.swift [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+  public var batchNorm2: BatchNorm<Float>
+
+  public init(filterCount: Int, pointwiseFilterCount: Int, depthMultiplier: Int,
+    strides: (Int, Int), printSummary: Bool = false) {
+    self.strides = strides
+    self.printSummary = printSummary
+    if depthMultiplier > 0 {
+      self.depthMultiplier = depthMultiplier
+    } else {
+      print("""
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Do we want to make this check a guard or precondition that triggers a fatalError when they specify an illegal value?

File: Models/ImageClassification/MobileNet.swift [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+  public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+    let convolved = input.sequenced(through: convBlock1, dConvBlock1,
+      dConvBlock2, dConvBlock3, dConvBlock4)
+    let convolved2 = convolved.sequenced(through: dConvBlock5, dConvBlock6,
+      dConvBlock7, dConvBlock8, dConvBlock9)
+    let convolved3 = convolved2.sequenced(through: dConvBlock10, dConvBlock11,
+      dConvBlock12, dConvBlock13)
+
+    // Verify shape since any discrepancies will be masked after pooling
+    let expectedWidth = input.shape[1] / 32
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: We typically haven't done shape checks inside the models themselves, and materializing the shape here may have adverse performance implications. If this is something we need to test against in unit tests, we may want to think about how we could extract that during a test there, rather than during normal execution.

File: Models/ImageClassification/MobileNet.swift [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      print("""
+            Invalid shape before GlobalAveragePooling2D: \(convolved2.shape)
+            should be [1, \(expectedWidth), \(expectedHeight), 1024].
+            """)
+    }
+
+    let convolved4 = convolved3.sequenced(through: avgPool, reshape,
+      dropoutLayer, convLast)
+    let output = convolved4.reshaped(to: [1, classCount])
+    if self.printSummary {
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Likewise, other models haven't employed shape logging as a standard operation. This might be something to verify in a unit test, but we probably don't want to have the logging internal to the model itself.

**Comment by BradLarson** [comment]
When it comes to linting, we use `swift-format` to format all of our files, and there is a `.swift-format` in this project that specifies our few format tweaks.

File: Models/ImageClassification/MobileNet.swift (hunk 7 above)
Comment (BradLarson) [reply Models/ImageClassification/MobileNet.swift]: Sorry, ignore the above comment. This is needed for proper autodiff.

File: Models/ImageClassification/MobileNet.swift (hunk 11 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: This was useful for verifying replication of the python implementation, but now that it's built that makes sense.

File: Models/ImageClassification/MobileNet.swift (hunk 10 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Removed from the execution path. I'll think about how to test for this externally.

File: Models/ImageClassification/MobileNet.swift (hunk 4 above)
Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Based on layer outputs and shapes, this seems correct. We can investigate further when Imagenette is complete and we have a working example.

File: Models/ImageClassification/MobileNet.swift [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            through: convBlock1, dConvBlock1,
+            dConvBlock2, dConvBlock3, dConvBlock4)
+        let convolved2 = convolved.sequenced(
+            through: dConvBlock5, dConvBlock6,
+            dConvBlock7, dConvBlock8, dConvBlock9)
+        let convolved3 = convolved2.sequenced(
+            through: dConvBlock10, dConvBlock11,
+            dConvBlock12, dConvBlock13)
+
+        // Verify shape since any discrepancies will be masked after pooling
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Because the code checking for this has been removed, this comment can probably be removed as well.

Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Good catch

File: Models/ImageClassification/MobileNet.swift [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    public var dConv: DepthwiseConv2D<Float>
+    public var batchNorm1: BatchNorm<Float>
+    public var conv: Conv2D<Float>
+    public var batchNorm2: BatchNorm<Float>
+
+    public init(
+        filterCount: Int, pointwiseFilterCount: Int, depthMultiplier: Int,
+        strides: (Int, Int)
+    ) {
+        if depthMultiplier < 1 {
```
Comment (BradLarson) [inline Models/ImageClassification/MobileNet.swift]: Rather than an if statement here, I believe a precondition would be preferred for an invalid argument value check like this. For example, we do [a similar check for random seeds](https://github.com/tensorflow/swift-apis/blob/master/Sources/Tensor/Random.swift#L125) in the swift-apis. That could condense this down to a single line.

Comment (texasmichelle) [reply Models/ImageClassification/MobileNet.swift]: Thanks for the example!
